core/src/main/kotlin/plugability/DokkaPlugin.kt

```diff
@@ -9,13 +9,25 @@ import kotlin.properties.ReadOnlyProperty
 import kotlin.reflect.KProperty
 import kotlin.reflect.KProperty1
 
+@RequiresOptIn(
+    level = RequiresOptIn.Level.WARNING,
+    message = "All of Dokka's plugin API is in preview and it can be changed " +
+            "in a backwards-incompatible manner with a best-effort migration. " +
+            "By opting in, you acknowledge the risks of relying on preview API."
+)
+@Target(AnnotationTarget.CLASS, AnnotationTarget.FUNCTION, AnnotationTarget.FIELD)
+@Retention(AnnotationRetention.BINARY)
+annotation class PreviewDokkaPluginApi
+
 abstract class DokkaPlugin {
     private val extensionDelegates = mutableListOf<KProperty<*>>()
     private val unsafePlugins = mutableListOf<Lazy<Extension<*, *, *>>>()
 
     @PublishedApi
     internal var context: DokkaContext? = null
 
+    @PreviewDokkaPluginApi
+    protected abstract fun previewDokkaPluginApiEmptyMethod()
```

Comment:
Some concerns with this method:

1. Whenever you override the method in IDEA, it propagates the annotation by default. You can even see it [below](https://github.com/Kotlin/dokka/pull/2743/files#diff-09b22991ead3a3b6b27039712b0391a89941978b85ae77f60016bf9ae39226ab) in one of our own plugins. So the user doesn't have to opt-in, and they don't even see the opt-in message.
2. The name doesn't reflect the intention, it's not very clear what this "empty method" is supposed to mean.
3. The return type can be confusing. Having to override a method that has an empty body feels wrong for some reason.
4. No documentation for the method. The opt-in annotation has the main message, but I think it's worth adding a simple explanation for why the method exists.

I like what Gradle did with Gradle Enterprise, where you have to set "yes" for `termsOfServiceAgree`:
https://github.com/Kotlin/dokka/blob/7544a215fb580ae0c47d1f397334f150d1a1ec65/settings.gradle.kts#L56-L62

It'd be nice to have something similar if we stick to the abstract function idea.

Maybe there's an obvious or just a better way to address the concerns I have (cc @qwwdfsad), but in the short time that I have I could only come up with the following:

```kotlin
enum class PluginApiPreviewAcknowledgement {
    @DokkaPluginApiPreview
    YES
}

abstract class DokkaPlugin {
    ...
    abstract fun pluginApiPreviewAcknowledgement(): PluginApiPreviewAcknowledgement
    ...
}
```

It could then be overridden in a plugin with

```kotlin
class GfmPlugin : DokkaPlugin() {
    ...
    @OptIn(DokkaPluginApiPreview::class)
    override fun pluginApiPreviewAcknowledgement(): PluginApiPreviewAcknowledgement {
        return PluginApiPreviewAcknowledgement.YES
    }
    ...
}
```

There's only one option to choose from, and when you use it - you have to opt in, which that makes you read the message, and it doesn't require to use the opt-in annotation in more than one place.

Reply:
As for me, I do not against it.

Can you clarify this point? I have not understood this case. What do you mean?
> So the user doesn't have to opt-in, and they don't even see the opt-in message.

Reply:
> Can you clarify this point? I have not understood this case. What do you mean?

You don't have to write `@OptIn(DokkaPluginApiPreview::class)` with the current approach, see [AndroidDocumentationPlugin](https://github.com/Kotlin/dokka/pull/2743/files#diff-09b22991ead3a3b6b27039712b0391a89941978b85ae77f60016bf9ae39226ab) or any other

Reply:
Nice catch with IDE propagation!

>Can you clarify this point? I have not understood this case. What do you mean?

Right now, if one implements `DokkaPlugin` and uses IDEA to autogenerate method bodies (alt+enter -> Implement members), IDEA automatically opt-ins:

```
class Foo : DokkaPlugin {
   
}
```

alt+enter -> enter -> enter ->

```
class Foo : DokkaPlugin {
    @DokkaPluginApiPreview // <- THIS LINE
   override fun previewDokkaPluginApiEmptyMethod() {}
}
```

and the user effectively does not receive opt-in message at all.

I am not really fond of additional steps with `pluginApiPreviewAcknowledgement = YES` (hello there, Gradle build scans), but it seems to be the only option.

TL;DR

* We have a preview annotation
* Abstract method itself is not marked with this annotation
* Method return type is singleton (`object`) with a meaningful name, so users don't any options except return this object immediately and explicitly opt-in.

The name of the method and object is subject to bikeshed, so pick any you like. The key here is to have a meaningful documentation (that mentions the purpose of the method, the fact it is not used and the example of implementation) and meaningful opt-in message

Reply:
(on a side note, we already have the concept of "enum with a single entry to emulate singleton", it's `object` :))

Reply:
Note: since it marks the whole function as `@RequiresOptIn`, we'll have to suppress it on our end so that we don't see the warning in our logs.

<details>
<summary>screenshot</summary>

![2023-02-13_16-40-33](https://user-images.githubusercontent.com/22685910/218503329-20c8a425-1f8a-414b-8fac-b69f2c21bd53.png)

</details>

Reply:
(just in case: I'm fine with both object and enum entry solutions, whichever one you prefer)

Reply:
In case of entry, IDE generetes :